framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java

```diff
@@ -569,30 +646,70 @@ public JavaExpression visit(NameExpr expr, JavaExpressionContext context) {
+         * @return a JavaExpression for the given name
+         */
+        private static FieldAccess getFieldJavaExpression(
+                VariableElement fieldElem,
+                JavaExpression receiverExpr,
+                boolean isOriginalReceiver) {
+            TypeMirror receiverType = receiverExpr.getType();
+
+            TypeMirror fieldType = ElementUtils.getType(fieldElem);
+            if (ElementUtils.isStatic(fieldElem)) {
+                Element classElem = fieldElem.getEnclosingElement();
+                JavaExpression staticClassReceiver = new ClassName(ElementUtils.getType(classElem));
+                return new FieldAccess(staticClassReceiver, fieldType, fieldElem);
+            }
+            JavaExpression locationOfField;
+            if (isOriginalReceiver) {
+                locationOfField = receiverExpr;
+            } else {
+                locationOfField = new ThisReference(receiverType);
+            }
+            if (locationOfField instanceof ClassName) {
+                throw new ParseRuntimeException(
+                        constructJavaExpressionParseError(
+                                fieldElem.getSimpleName().toString(),
+                                "a non-static field cannot have a class name as a receiver."));
+            }
+            return new FieldAccess(locationOfField, fieldType, fieldElem);
+        }
+
         @Override
-        public JavaExpression visit(MethodCallExpr expr, JavaExpressionContext context) {
+        public JavaExpression visit(MethodCallExpr expr, Void aVoid) {
             setResolverField();
 
             JavaExpression receiverExpr;
             if (expr.getScope().isPresent()) {
-                receiverExpr = expr.getScope().get().accept(this, context);
+                receiverExpr = expr.getScope().get().accept(this, null);
                 expr = expr.removeScope();
+            } else if (thisReference != null) {
+                receiverExpr = thisReference;
             } else {
-                receiverExpr = context.receiver;
+                receiverExpr = new ClassName(enclosingType);
             }
 
             String methodName = expr.getNameAsString();
 
             // parse argument list
-            List<JavaExpression> arguments = new ArrayList<>();
-            if (!expr.getArguments().isEmpty()) {
-                JavaExpressionContext argContext = context;
-                for (Expression argument : expr.getArguments()) {
-                    arguments.add(argument.accept(this, argContext));
-                }
+            List<JavaExpression> arguments = new ArrayList<>(expr.getArguments().size());
+            for (Expression argument : expr.getArguments()) {
+                arguments.add(argument.accept(this, null));
             }
 
-            // Find the method element.
             ExecutableElement methodElement;
             try {
                 methodElement =
```

Comment:
This call to `getMethodElement` does not seem to account for autoboxing.  (If it does, that should be documented in `getMethodElement`.  Therefore, I am concerned whether the (right) method will be found.

Also, why does `getMethodElement` return `Element` rather than `ExecutableElement`?  I suggest it be changed, which permits some local variable types to be changed and casts eliminated.  Within its implementation, how can the test `(element.getKind() == ElementKind.METHOD)` evaluate to false?  If it does somehow, should `element` be reset to `null`?

I suggest making these changes in a separate pull request because they are not directly related to the current pull request.

Reply:
I've done so here: https://github.com/typetools/checker-framework/pull/4394.